File: src/main/java/org/frawa/elmtest/core/ElmProjectTestsHelper.java
package org.frawa.elmtest.core;

import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.project.Project;
import org.elm.workspace.ElmProject;
import org.elm.workspace.ElmWorkspaceService;
import org.jetbrains.annotations.NotNull;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Optional;
import java.util.stream.Stream;

public class ElmProjectTestsHelper {

    private final ElmWorkspaceService workspaceService;

    public ElmProjectTestsHelper(Project project) {
        workspaceService = ServiceManager.getService(project, ElmWorkspaceService.class);
    }

    public Stream<String> allNames() {
        return getTestableProjects()
                .map(ElmProject::getPresentableName);
    }

    @NotNull
    private Stream<ElmProject> getTestableProjects() {
        return workspaceService.getAllProjects().stream()
                .filter(p -> Files.exists(p.getProjectDirPath().resolve("tests")));
    }

    public Optional<String> nameByProjectDirPath(String path) {
        return getTestableProjects()
                .filter(p -> p.getProjectDirPath().toString().equals(path))
                .map(ElmProject::getPresentableName)
                .findFirst();
    }

    public Optional<String> projectDirPathByName(String name) {
        return getTestableProjects()
                .filter(p -> p.getPresentableName().equals(name))
                .map(ElmProject::getProjectDirPath)
                .map(Path::toString)
                .findFirst();
    }

    public Optional<ElmProject> elmProjectByProjectDirPath(String path) {
        return getTestableProjects()
                .filter(p -> p.getProjectDirPath().toString().equals(path))
                .findFirst();
    }

    public Path adjustElmCompilerProjectDirPath(String elmFolder, Path compilerPath) {
        return elmProjectByProjectDirPath(elmFolder)
                .filter(ElmProject::isElm18)
                .map(_1 -> compilerPath.resolveSibling("elm-make"))
                .orElse(compilerPath);
    }

    public static Path elmFolderForTesting(ElmProject elmProject) {
        return elmProject.isElm18() && elmProject.getPresentableName().equals("tests")
                ? elmProject.getProjectDirPath().getParent()
                : elmProject.getProjectDirPath();
    }

}

File: src/main/kotlin/org/elm/ide/test/core/ElmProjectTestsHelper.kt
package org.elm.ide.test.core

import com.intellij.openapi.project.Project
import com.intellij.util.io.exists
import org.elm.workspace.ElmProject
import org.elm.workspace.elmWorkspace

class ElmProjectTestsHelper(project: Project) {

    private val elmWorkspace = project.elmWorkspace

    private val testableProjects: List<ElmProject>
        get() = elmWorkspace.allProjects
                .filter { it.testsDirPath.exists() }

    fun allNames() = testableProjects.map { it.presentableName }

    fun nameByProjectDirPath(path: String): String? {
        return testableProjects
                .filter { it.projectDirPath.toString() == path }
                .map { it.presentableName }
                .firstOrNull()
    }

    fun projectDirPathByName(name: String): String? {
        return testableProjects
                .filter { it.presentableName == name }
                .map { it.projectDirPath.toString() }
                .firstOrNull()
    }

    fun elmProjectByProjectDirPath(path: String): ElmProject? {
        return testableProjects
                .firstOrNull { it.projectDirPath.toString() == path }
    }
}
